File: campaign/campaign.go
package campaign

import (
	"time"

	"thyago.com/otelinho/targetingrules"
)

type Campaign struct {
	ID              int
	Creative        string
	StartDate       time.Time
	EndDate         time.Time
	Goal            uint
	MaxBid          float64
	RemainingBudget float64
	Budget          float64
	PacingFactor    int64
	Targeting       []targetingrules.TargetingRule
}

File: lib/otelinho_ad_server/campaign.ex
defmodule OtelinhoAdServer.Campaign do
  use Ecto.Schema
  import Ecto.Changeset

  schema "campaigns" do
    field :budget, :decimal
    field :creative, :string
    field :start_date, :utc_datetime
    field :end_date, :utc_datetime
    field :goal, :integer
    field :max_bid, :decimal
    field :remaining_budget, :decimal
  end

  @doc false
  def changeset(campaign, attrs) do
    campaign
    |> cast(attrs, [:creative, :start_date, :end_date, :goal, :max_bid, :budget, :remaining_budget])
    |> validate_required([:creative, :start_date, :end_date, :goal, :max_bid, :budget, :remaining_budget])
  end
end
